### ISSUE TITLE
PDF content has extra spaces inserted into it

### ISSUE THREAD

**Issue by LordMonoxide** [smalot/pdfparser#396]
My company uses `wkhtmltopdf` to generate PDF files. Our automated testing uses `pdfparser` to verify that the generated content matches what we expect. I'm in the process of updating from `v0.11` to `v0.18.2`. We got the expected output in the older version, but the newer version is inserting extra spaces into the content.

**Expected content:** Snippet 1 EN
**Actual content:** S nip pet 1  E N

Note that I have stepped through the code and verified that they are in fact spaces and not other bad unicode characters, and found the code that is inserting the spaces.

I'm not sure of the actual terminology here, since I'm unfamiliar with the PDF format, but I'll do my best to describe what's going on.

The actual "content" of the PDF file is as follows:
```
/F6 16 Tf 1 0 0 -1 0 0 Tm
8 -15 Td <0001> Tj
10.1562500 0 Td <0002> Tj
10.1406250 0 Td <0003> Tj
4.43750000 0 Td <0004> Tj
10.1562500 0 Td <0004> Tj
10.1562500 0 Td <0005> Tj
9.84375000 0 Td <0006> Tj
6.26562500 0 Td <0007> Tj
5.07812500 0 Td <0008> Tj
10.1718750 0 Td <0007> Tj
5.07812500 0 Td <0009> Tj
10.1093750 0 Td <000A> Tj 
```

From what I understand, `Td` is an `x, y` offset, and `Tj` is an index into an encoding table, as follows:
```
0001 = 'S'
0002 = 'n'
0003 = 'i'
0004 = 'p'
0005 = 'e'
0006 = 't'
0007 = ' '
0008 = '1'
0009 = 'E'
000A = 'N'
```

As you can see, the content is correct and does not contain extra spaces. The issue seems to come from the `Td` command. Some have wider `x` offsets, causing the following code in `pdfparser` to insert extra spaces (starting at `PDFObject:289`, "`// horizontal offset`"):
```
                    // move text current point
                    case 'Td':
                        $args = preg_split('/\s/s', $command[self::COMMAND]);
                        $y = array_pop($args);
                        $x = array_pop($args);
                        if (((float) $x <= 0) ||
                            (false !== $current_position_td['y'] && (float) $y < (float) ($current_position_td['y']))
                        ) {
                            // vertical offset
                            $text .= "\n";
                        } elseif (false !== $current_position_td['x'] && (float) $x > (float) (
                                $current_position_td['x']
                            )
                        ) {
                            // horizontal offset
                            $text .= ' ';
                        }
                        $current_position_td = ['x' => $x, 'y' => $y];
                        break;
```

I have tested the exact file in other parsers (gimp, okular, chrome, firefox, edge, word) and all render correctly.

I have attached the file for reference.

[36360.00.d28.pdf](https://github.com/smalot/pdfparser/files/6045453/36360.00.d28.pdf)

**Comment by LordMonoxide** [comment]
Apologies, I just realized we had `$text .= ' ';` commented out in the vendor code when we were using `v0.11`. I reverted to the original `v0.11` and it is indeed broken in that version, too.

**Comment by k00ni** [comment]
Thank you for your detailed report!

Can we include your PDF in our tests? Is it free of charge and without any obligations?

@Connum @izabala, any idea?

**Comment by LordMonoxide** [comment]
Absolutely, feel free to use anything I've included in the issue

**Comment by BugHunter2k** [comment]
I'm seeing this too. 
It seems that all pdf-files generated by wkhtmltopdf are affected.

I'll attach one testfile.
[test.pdf](https://github.com/smalot/pdfparser/files/6130367/test.pdf)

**Comment by k00ni** [comment]
CC @smalot

**Comment by harrisnl** [comment]
I'm also experiencing this issue when converting pdf files.
Would be great if it can be fixed 👍

**Comment by kaduhoff** [comment]
I did a test on my application, I changed the line 410 in the class Font, and everything seems fine. No more spaces between the words.  
original
`return implode(' ', $words);`

new one
`return implode('', $words);`

My version is 0.18.2
I hope that you lanch a new release soon. Thanks

**Comment by Connum** [comment]
> I did a test on my application, I changed the line 410 in the class Font, and everything seems fine. No more spaces between the words.
> 
> Smalot\PdfParser\Encoding\WinAnsiEncoding\Font
> original
> `return implode(' ', $words);`
> 
> new one
> `return implode('', $words);`

I don't have the time to look into this right now, unfortunately. Just dropped in to say that this might possibly break cases where the spacing is currently fine. It would be better to get to the underlying cause of the different spacing, and either that can be addressed directly, or it could be detected and then a switch added to either implode the words with or without a space.

**Comment by BugHunter2k** [comment]
> I did a test on my application, I changed the line 410 in the class Font, and everything seems fine. No more spaces between the words.
> 
> Smalot\PdfParser\Encoding\WinAnsiEncoding\Font
> original
> `return implode(' ', $words);`
> 
> new one
> `return implode('', $words);`
> 
> My version is 0.18.2
> I hope that you lanch a new release soon. Thanks

Chaned the file and tried this some of my files (including the test.pdf).
I see less spaces but there are still some additional spaces inserted. So this seems not to be the solution.

**Comment by LordMonoxide** [comment]
FWIW, this is the vendor patch that I'm currently using. Note that this works **in my case** and may break legitimate spacing in other cases.

```
--- ./vendor/smalot/pdfparser/src/Smalot/PdfParser/PDFObject.php	2021-02-25 03:40:30.000000000 -0400
+++ ./vendor/smalot/pdfparser/src/Smalot/PdfParser/PDFObject.patched.php	2021-02-25 16:45:49.700000000 -0400
@@ -300,7 +300,9 @@
                             )
                         ) {
                             // horizontal offset
-                            $text .= ' ';
+                            //TODO PATCH:
+                            // See https://github.com/smalot/pdfparser/issues/396
+                            //$text .= ' ';
                         }
                         $current_position_td = ['x' => $x, 'y' => $y];
                         break;
```

**Comment by BugHunter2k** [comment]
> FWIW, this is the vendor patch that I'm currently using. Note that this works **in my case** and may break legitimate spacing in other cases.

Tried this and it does **not** work for my test.pdf

**Comment by LucianoHanna** [comment]
I made some debugging here, looks like this case causes the bug (function getText at PDFObject:288), when I remove this, the getText return the right string
```php
// move text current point
case 'Td':
    $args = preg_split('/\s/s', $command[self::COMMAND]);
    $y = array_pop($args);
    $x = array_pop($args);
    if (((float) $x <= 0) ||
        (false !== $current_position_td['y'] && (float) $y < (float) ($current_position_td['y']))
    ) {
        // vertical offset
        $text .= "\n";
    } elseif (false !== $current_position_td['x'] && (float) $x > (float) (
        $current_position_td['x']
    )
    ) {
        // horizontal offset
        $text .= ' ';
    }
$current_position_td = ['x' => $x, 'y' => $y];
break;
```

**Comment by BugHunter2k** [comment]
> I made some debugging here, looks like this case causes the bug (function getText at PDFObject:288), when I remove this, the getText return the right string

Tried this with multiple files. It worked with every single one. 
So this is where the spaces come from.

**Comment by Lucasthoelke** [comment]
Any status update on this?

**Comment by nicholaszuccarelli** [comment]
Sorry to be a pain but figured I'd bump this. Having the same issue with Wkhtmltopdf documents including odd whitespaces which is throwing my document off.

**Comment by GeoNomad** [comment]
New here and the first document I tried pdfparser on threw in extra spaces in the middle of a few words. They don't appear in renderings of the PDF I have tried. Maybe invisible characters in the original, which seems to have been created by Word 2013 according to getDetails.

to see k out 
the time they join ed
need w on’t arise

Original document is here: https://probusglobal.org/Newsletters/2021-09_Holland_Park.pdf

I don't know if this relates to previous comments, or is a different issue.

**Comment by chrisbaswell** [comment]
Hey I'm trying to parse a document created with Wkhtmltopdf to get the x and y coordinates of a tag like [signature|signer1] and ran into this issue.  Every single character on the document is getting parsed as a individual piece of text so when calling $page->getDataTim() for instance I receive an array for each character in my tag - `[["1","0","0","-1","348","-250"],"["],[["1","0","0","-1","353.53125","-250"],"s"],...`.  Considering I just need the coordinates of my tag and don't care about the text itself, is there a quick and dirty fix that I could try to get around this issue?

**Comment by ivordurham** [comment]
Not sure if I have stumbled onto the same problem or a related one. I have a Word file with just "This is a test!" as the content and saved it as a PDF. The getText method returns "Th\tis is a test!" (i.e. a tab splits the first word.) I've attached the original Word document and the PDF it saved. Copying/pasting from Acrobat produces the correct plain text.
[Essay Test.pdf](https://github.com/smalot/pdfparser/files/7457955/Essay.Test.pdf)
[Essay Test.docx](https://github.com/smalot/pdfparser/files/7457956/Essay.Test.docx)
d.

**Comment by rubenvanerk** [comment]
Using #505, this seems to fix the issue:

```php
$config = new \Smalot\PdfParser\Config();
$config->setHorizontalOffset('');

$parser = new \Smalot\PdfParser\Parser([], $config);
$pdf    = $parser->parseFile('document.pdf');
 
$text = $pdf->getText('path/to/36360.00.d28.pdf');
```

Outputs:
```text
Snippet 1 EN
```

**Comment by k00ni** [comment]
I merged #505 into `master` branch. It will be part of the next release.

**Comment by stoyanov-sv** [comment]
Faced a similar problem. It seems to me that it appears when the pdf is of a higher version than 1.4.
If you downgrade the pdf with ghostscript it works. As a temporary solution.

`gs -dCompatibilityLevel=1.4 -dBATCH -dNOPAUSE -q -sDEVICE=pdfwrite -sOutputFile=new.pdf origin.pdf`
